snuba/datasets/entities/factory.py

```diff
@@ -1,89 +1,135 @@
-    from snuba.datasets.entities.generic_metrics import (
-        GenericMetricsDistributionsEntity,
-        GenericMetricsSetsEntity,
-    )
-    from snuba.datasets.entities.metrics import (
-        MetricsCountersEntity,
-        MetricsDistributionsEntity,
-        MetricsSetsEntity,
-        OrgMetricsCountersEntity,
-    )
-    from snuba.datasets.entities.outcomes import OutcomesEntity
-    from snuba.datasets.entities.outcomes_raw import OutcomesRawEntity
-    from snuba.datasets.entities.profiles import ProfilesEntity
-    from snuba.datasets.entities.replays import ReplaysEntity
-    from snuba.datasets.entities.sessions import OrgSessionsEntity, SessionsEntity
-    from snuba.datasets.entities.transactions import TransactionsEntity
-
-    dev_entity_factories: MutableMapping[EntityKey, Callable[[], Entity]] = {}
-
-    entity_factories: MutableMapping[EntityKey, Callable[[], Entity]] = {
-        EntityKey.DISCOVER: DiscoverEntity,
-        EntityKey.EVENTS: EventsEntity,
-        EntityKey.GROUPASSIGNEE: GroupAssigneeEntity,
-        EntityKey.GROUPEDMESSAGES: GroupedMessageEntity,
-        EntityKey.OUTCOMES: OutcomesEntity,
-        EntityKey.OUTCOMES_RAW: OutcomesRawEntity,
-        EntityKey.SESSIONS: SessionsEntity,
-        EntityKey.ORG_SESSIONS: OrgSessionsEntity,
-        EntityKey.TRANSACTIONS: TransactionsEntity,
-        EntityKey.DISCOVER_TRANSACTIONS: DiscoverTransactionsEntity,
-        EntityKey.DISCOVER_EVENTS: DiscoverEventsEntity,
-        EntityKey.METRICS_SETS: MetricsSetsEntity,
-        EntityKey.METRICS_COUNTERS: MetricsCountersEntity,
-        EntityKey.ORG_METRICS_COUNTERS: OrgMetricsCountersEntity,
-        EntityKey.METRICS_DISTRIBUTIONS: MetricsDistributionsEntity,
-        EntityKey.PROFILES: ProfilesEntity,
-        EntityKey.FUNCTIONS: FunctionsEntity,
-        EntityKey.REPLAYS: ReplaysEntity,
-        EntityKey.GENERIC_METRICS_SETS: GenericMetricsSetsEntity,
-        EntityKey.GENERIC_METRICS_DISTRIBUTIONS: GenericMetricsDistributionsEntity,
-        **(dev_entity_factories if settings.ENABLE_DEV_FEATURES else {}),
-    }
-
-    try:
-        entity = ENTITY_IMPL[name] = entity_factories[name]()
-        ENTITY_NAME_LOOKUP[entity] = name
-    except KeyError as error:
-        raise InvalidEntityError(f"entity {name!r} does not exist") from error
-
-    return entity
+    return _ent_factory().get(name)
+
+
+def get_entity_name(entity: Entity) -> EntityKey:
+    return _ent_factory().get_entity_name(entity)
+
+
+def get_all_entity_names() -> Sequence[EntityKey]:
+    return _ent_factory().all_names()
 
 
 def enforce_table_writer(entity: Entity) -> TableWriter:
     writable_storage = entity.get_writable_storage()
 
     assert (
         writable_storage is not None
-    ), f"Entity {ENTITY_NAME_LOOKUP[entity]} does not have a writable storage."
+    ), f"Entity {_ent_factory().get_entity_name(entity)} does not have a writable storage."
     return writable_storage.get_table_writer()
+
```

Comment:
Needed to add entity map override functionality to keep some test cases happy (ref. https://github.com/getsentry/snuba/blob/df9871f31d88da6e2d01ee7b0422280a63fa6b6e/tests/query/joins/test_equivalence_adder.py#L261)